content/concepts/consensus-network/invariant-checking.md

```diff
@@ -0,0 +1,404 @@
+- [XRP Not Created](#xrp-not-created)
+- [Account Roots Not Deleted](#account-roots-not-deleted)
+- [XRP Balance Checks](#xrp-balance-checks)
+- [Ledger Entry Types Match](#ledger-entry-types-match)
+- [No XRP Trust Lines](#no-xrp-trust-lines)
+- [No Bad Offers](#no-bad-offers)
+- [No Zero Escrow](#no-zero-escrow)
+- [Valid New Account Root](#valid-new-account-root)
+
+
+### Transaction Fee Check
+
+- **Condition(s) Checked:**
+
+    - [Trnsaction fees](https://xrpl.org/rippleapi-reference.html#transaction-fees) should never be negative nor larger than the transaction itself.
+
+```
+class TransactionFeeCheck
+{
+public:
+    void
+    visitEntry(
+        bool,
+        std::shared_ptr<SLE const> const&,
+        std::shared_ptr<SLE const> const&);
+
+    bool
+    finalize(
+        STTx const&,
+        TER const,
+        XRPAmount const,
+        ReadView const&,
+        beast::Journal const&);
+};
+```
+
+### XRP Not Created
+
+- **Condition(s) Checked:**
+    
+    - A transaction must not create XRP and should only destroy the XRP [fee](https://xrpl.org/rippleapi-reference.html#transaction-fees)).
+
+```
+class XRPNotCreated
+{
+    std::int64_t drops_ = 0;
+
+public:
+    void
+    visitEntry(
+        bool,
+        std::shared_ptr<SLE const> const&,
+        std::shared_ptr<SLE const> const&);
+
+    bool
+    finalize(
+        STTx const&,
+        TER const,
+        XRPAmount const,
+        ReadView const&,
+        beast::Journal const&);
+};
+```
+
+### Account Roots Not Deleted
+
+- **Condition(s) Checked:**
+    
+    - An account ledger entry cannot be removed.
+
+```
+class AccountRootsNotDeleted
+{
+    std::uint32_t accountsDeleted_ = 0;
+
+public:
+    void
+    visitEntry(
+        bool,
+        std::shared_ptr<SLE const> const&,
+        std::shared_ptr<SLE const> const&);
+
+    bool
+    finalize(
+        STTx const&,
+        TER const,
+        XRPAmount const,
+        ReadView const&,
+        beast::Journal const&);
+};
+```
+
+### XRP Balance Checks
+
+- **Condition(s) Checked:**
+    
+    - An account's XRP balance must be of type XRP and take a value between 0 and INITIAL_XRP drops, inclusive.
+
+```
+/** Number of drops per 1 XRP */
+constexpr XRPAmount DROPS_PER_XRP{1'000'000};
+
+/** Number of drops in the genesis account. */
+constexpr XRPAmount INITIAL_XRP{100'000'000'000 * DROPS_PER_XRP};
+```
+
+```
+class XRPBalanceChecks
+{
+    bool bad_ = false;
+
+public:
+    void
+    visitEntry(
+        bool,
+        std::shared_ptr<SLE const> const&,
+        std::shared_ptr<SLE const> const&);
+
+    bool
+    finalize(
+        STTx const&,
+        TER const,
+        XRPAmount const,
+        ReadView const&,
+        beast::Journal const&);
+};
+```
+
+### Ledger Entry Types Match
+
+- **Condition(s) Chceked:**
+    
+    - Corresponding modified ledger entries should match in type and added entries should be a [valid type](https://xrpl.org/transaction-types.html#transaction-types).
```

Comment:
Wrong link. "Ledger entry types" are the same thing as ["ledger object types"](https://xrpl.org/ledger-object-types.html). Also, this would make more sense if there were some more context about what's "corresponding" to what in this one. I think it means the objects that the transaction modified, as identified by their IDs, before and after executing the transaction? Not sure if it's exactly looking at the [metadata](https://xrpl.org/transaction-metadata.html).